hibernate-core/src/main/java/org/hibernate/query/internal/ParameterMetadataImpl.java

```diff
@@ -42,19 +43,19 @@ public class ParameterMetadataImpl implements ParameterMetadataImplementor {
 	public static final ParameterMetadataImpl EMPTY = new ParameterMetadataImpl();
 
 	private final Map<QueryParameterImplementor<?>, List<SqmParameter<?>>> queryParameters;
-
-	private final Set<String> names;
-	private final Set<Integer> labels;
-
+	private final Map<String, QueryParameterImplementor<?>> queryParametersByNames;
+	private final Map<Integer, QueryParameterImplementor<?>> queryParametersByLabels;
 
 	private ParameterMetadataImpl() {
 		this.queryParameters = Collections.emptyMap();
-		this.names = Collections.emptySet();
-		this.labels = Collections.emptySet();
+		this.queryParametersByNames = Collections.emptyMap();
+		this.queryParametersByLabels = Collections.emptyMap();
 	}
 
 	public ParameterMetadataImpl(Map<QueryParameterImplementor<?>, List<SqmParameter<?>>> queryParameters) {
 		this.queryParameters = queryParameters;
+		this.queryParametersByNames = new HashMap<>();
+		this.queryParametersByLabels = new HashMap<>();
```

Comment:
Might be worth only initializing the maps if needed, i.e. if there are named / positional parameters as most of the time neither or only 1 type of parameters is used IMO. Not that big of a deal if it's easier this way.

Reply:
I agree: we can trust that in the majority of use cases, people will be using either named or ordered parameters.

Which implies only one of these `HashMap` instances will be actually used - and since these are expensive to allocate, I'd also prefer to avoid it.

Reply:
P.S. would it be possible to use an array instead of a `Map<Integer, value > ` ?

I think that would be sensible here, as the numbers aren't going to have gaps. The only drawback is that implementation-wise, you'll need to figure out the maximum size of the array before being able to start collecting them; this might imply needing an additional iteration within the constructor, but it's a very fast scan - and you can avoid doing it if there's no positional parameters.

Reply:
another problem using an array can be the method `Set<Integer> getOrdinalParameterLabels()`,

Reply:
ok, feel free to ignore my remark about Maps and Sets of Integer, we can look into that separately.

Reply:
anyway I made a change to initialize them only if necessary